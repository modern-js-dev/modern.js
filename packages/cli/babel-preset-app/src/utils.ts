import fs from 'fs';
import path from 'path';
import semver from 'semver';

export const isBeyondReact17 = (cwd: string) => {
  const pkgInfo = JSON.parse(
    fs.readFileSync(path.join(cwd, 'package.json'), 'utf8'),
  );

  const deps = { ...pkgInfo.devDependencies, ...pkgInfo.dependencies };

  if (typeof deps.react !== 'string') {
    return false;
  }

  return semver.satisfies(semver.minVersion(deps.react)!, '>=17.0.0');
};

// export const isPnpm = (appDirectory: string) => {
//   // pnpm 以及 rush 场景需要编译
//   const currentDir = appDirectory || process.cwd();
//   if (findMonorepoRoot(currentDir)) {
//     return getPkgManager(currentDir) === 'pnpm';
//   } else if (
//     fs.existsSync(path.resolve(currentDir, './pnpm-lock.yaml')) ||
//     fs.existsSync(path.resolve(currentDir, './.rush'))
//   ) {
//     return true;
//   }

//   return false;
// };
