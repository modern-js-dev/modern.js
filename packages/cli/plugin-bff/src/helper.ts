import Module from 'module';
import * as path from 'path';

const serverRuntimeAlias = '@modern-js/runtime/server';
const serverRuntimePath = '.runtime-exports/server';

const registerModernRuntimePath = (internalDirectory: string) => {
  const originalResolveFilename = (Module as any)._resolveFilename;
  // eslint-disable-next-line node/no-unsupported-features/node-builtins
  const { builtinModules } = Module;
  (Module as any)._resolveFilename = function (
    request: string,
    _parent: any,
  ): string {
    const isCoreModule = builtinModules.includes(request);
    if (!isCoreModule) {
      if (request === serverRuntimeAlias) {
        const found = path.join(internalDirectory, serverRuntimePath);
        if (found) {
          // eslint-disable-next-line prefer-rest-params
          const modifiedArguments = [found, ...[].slice.call(arguments, 1)]; // Passes all arguments. Even those that is not specified above.
          return originalResolveFilename.apply(this, modifiedArguments);
        }
      }
    }
    // eslint-disable-next-line prefer-rest-params
    return originalResolveFilename.apply(this, arguments);
  };

  return () => {
    (Module as any)._resolveFilename = originalResolveFilename;
  };
};

export { registerModernRuntimePath };
