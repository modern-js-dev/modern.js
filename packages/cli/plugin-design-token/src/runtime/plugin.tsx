import type { Plugin } from '@modern-js/runtime-core';
import React, { useContext } from 'react';

export const DesignTokenContext = React.createContext<any>({});

export const useDesignTokens = (): IDesignTokens =>
  useContext<IDesignTokens>(DesignTokenContext);

export default (
  options: {
    token?: Record<string, any>;
    useStyledComponentsThemeProvider?: boolean;
    useDesignTokenContext?: boolean;
  } = {},
): Plugin => ({
  name: '@modern-js/plugin-design-token',

  setup: () => ({
    hoc({ App }, next) {
      return next({
        App: (props: any) => {
          const {
            token = {},
            useStyledComponentsThemeProvider = false,
            useDesignTokenContext = false,
          } = options;
          if (useStyledComponentsThemeProvider && useDesignTokenContext) {
            const { ThemeProvider } = require('@modern-js/runtime-core/styled');
            ThemeProvider.init =
              (App as React.ComponentType<any> & { init: () => void }).init ||
              // eslint-disable-next-line @typescript-eslint/no-empty-function
              function () {};
            return (
              <ThemeProvider theme={token}>
                <DesignTokenContext.Provider value={token}>
                  <App {...props} />
                </DesignTokenContext.Provider>
              </ThemeProvider>
            );
          } else if (useStyledComponentsThemeProvider) {
            const { ThemeProvider } = require('@modern-js/runtime-core/styled');
            ThemeProvider.init =
              (App as React.ComponentType<any> & { init: () => void }).init ||
              // eslint-disable-next-line @typescript-eslint/no-empty-function
              function () {};
            return (
              <ThemeProvider theme={token}>
                <App {...props} />
              </ThemeProvider>
            );
          } else if (useDesignTokenContext) {
            (
              DesignTokenContext as React.Context<any> & { init: () => void }
            ).init =
              (App as React.ComponentType<any> & { init: () => void }).init ||
              // eslint-disable-next-line @typescript-eslint/no-empty-function
              function () {};
            return (
              <DesignTokenContext.Provider value={token}>
                <App {...props} />
              </DesignTokenContext.Provider>
            );
          } else {
            return <App {...props} />;
          }
        },
      });
    },
  }),
});
