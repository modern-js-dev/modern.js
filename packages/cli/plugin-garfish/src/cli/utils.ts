import type { NormalizedConfig } from '@modern-js/core';

export const makeProvider = () => `
export const provider = function ({basename, dom}) {
  return {
    render({basename, dom, props}) {
      const SubApp = render({ props, basename });
      const node = dom.querySelector('#' + MOUNT_ID) || dom;
      const App = function () {
        return React.createElement(SubApp, props)
      };
      bootstrap(App, node);
    },
    destroy({ dom }) {
      const node = dom.querySelector('#' + MOUNT_ID) || dom;

      if (node) {
        unmountComponentAtNode(node);
      }
    },
    SubModuleComponent: (props) => {
      const SubApp = render({props, basename});

      return createPortal(<SubApp />, dom.querySelector('#' + MOUNT_ID)  || dom);
    },
    jupiter_submodule_app_key: (props) => {
      const SubApp = render({props, basename});

      return createPortal(<SubApp />, dom.querySelector('#' + MOUNT_ID)  || dom);
    }
  }
};

if (typeof __GARFISH_EXPORTS__ !== 'undefined') {
  __GARFISH_EXPORTS__.provider = provider;
}
`;

export const makeRenderFunction = (code: string) => {
  const inGarfishToRender = `
  const { basename, props } = arguments[0] || {};
  let renderByGarfish = false;
  const renderByProvider = !!basename;

  if (IS_BROWSER && window.Garfish && window.Garfish.activeApps && window.Garfish.activeApps.length !== 0) renderByGarfish = true;
  if (IS_BROWSER && window.Garfish && window.Garfish.apps && Object.keys(window.Garfish.apps).length !== 0) renderByGarfish = true;
  if (typeof __GARFISH_EXPORTS__ !== 'undefined') renderByGarfish = true;
  if (renderByGarfish && !renderByProvider) return null;

  function RouterPlugin (routerConfig) {
    if (basename) {
      routerConfig.basename = basename;
      if (routerConfig.supportHtml5History !== false) {
        if (!routerConfig.historyOptions) {
          routerConfig.historyOptions = {
            basename: basename
          };
        } else {
          routerConfig.historyOptions.basename = basename;
        }
      }
    }
    return router(routerConfig);
  }
  `;
  return (
    inGarfishToRender +
    code
      .replace(`router(`, `RouterPlugin(`)
      .replace('IS_BROWSER', `IS_BROWSER && !renderByGarfish`)
  );
};

// support legacy config
export function getRuntimeConfig(config: Partial<NormalizedConfig>) {
  if (config?.runtime?.features) {
    return config?.runtime?.features;
  }
  return config?.runtime || {};
}

// support legacy config
export function setRuntimeConfig(
  config: Partial<NormalizedConfig>,
  key: string,
  value: any,
): undefined {
  if (config?.runtime?.features && config?.runtime?.features[key]) {
    config.runtime.features[key] = value;
    return undefined;
  }
  if (config?.runtime && config?.runtime[key]) {
    config.runtime[key] = value;
    return undefined;
  }
  return undefined;
}
