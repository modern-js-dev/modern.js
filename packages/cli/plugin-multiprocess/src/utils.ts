export const getRequireCode = (plugins: string[]) =>
  plugins
    .map(
      (p, index) => `const mod_${index} = require('${p}');
const plugin_${index} =  mod_${index}.default || mod_${index};`,
    )
    .join('\n');

export const getPluginsCode = (plugins: string[]) =>
  `[${plugins.map((p, index) => `plugin_${index}`).join(',')}]`;
