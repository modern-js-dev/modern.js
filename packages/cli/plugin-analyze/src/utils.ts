import fs from 'fs';
import { path, INTERNAL_DIR_ALAIS, INTERNAL_SRC_ALIAS } from '@modern-js/utils';
import type { Entrypoint } from '@modern-js/types';
import type { ImportStatement } from './generateCode';
import { FILE_SYSTEM_ROUTES_FILE_NAME } from './constants';

export const walkDirectory = (dir: string): string[] =>
  fs.readdirSync(dir).reduce<string[]>((previous, filename) => {
    const filePath = path.join(dir, filename);
    if (fs.statSync(filePath).isDirectory()) {
      return [...previous, ...walkDirectory(filePath)];
    } else {
      return [...previous, filePath];
    }
  }, []);

export const getDefaultImports = ({
  entrypoint,
  srcDirectory,
}: {
  entrypoint: Entrypoint;
  srcDirectory: string;
}): ImportStatement[] => {
  const { entryName, fileSystemRoutes, customBootstrap, entry } = entrypoint;
  const imports = [
    {
      specifiers: [{ local: 'React' }],
      value: 'react',
    },
    {
      specifiers: [{ imported: 'createApp' }, { imported: 'bootstrap' }],
      value: '@modern-js/runtime',
    },
    customBootstrap && {
      specifiers: [{ local: 'customBootstrap' }],
      value: customBootstrap.replace(srcDirectory, INTERNAL_SRC_ALIAS),
    },
  ].filter(Boolean) as ImportStatement[];

  if (fileSystemRoutes) {
    const route: ImportStatement = {
      specifiers: [{ imported: 'routes' }],
      value: `${INTERNAL_DIR_ALAIS}/${entryName}/${FILE_SYSTEM_ROUTES_FILE_NAME}`,
    };
    if (fileSystemRoutes.globalApp) {
      imports.push({
        specifiers: [{ local: 'App' }],
        value: fileSystemRoutes.globalApp.replace(
          srcDirectory,
          INTERNAL_SRC_ALIAS,
        ),
      });
    } else {
      route.initialize = 'const App = false;';
    }

    imports.push(route);
  } else {
    imports.push({
      specifiers: [{ local: 'App' }],
      value: entry.replace(srcDirectory, INTERNAL_SRC_ALIAS),
    });
  }

  return imports;
};
