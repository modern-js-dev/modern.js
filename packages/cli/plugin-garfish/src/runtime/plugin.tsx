import { createPlugin } from '@modern-js/runtime-core';
import React from 'react';
import hoistNonReactStatics from 'hoist-non-react-statics';
import { logger } from '../util';
import { GarfishProvider } from './utils/Context';
import setExternal from './utils/setExternal';
import {
  Config,
  Manifest,
  MicroComponentProps,
  ModulesInfo,
  Options,
} from './useModuleApps';
import { generateMApp } from './utils/MApp';
import { AppMap, generateApps } from './utils/apps';

async function initOptions(manifest: Manifest = {}, options: Options) {
  let apps: ModulesInfo = [];

  // use manifest modules
  if (manifest?.modules) {
    apps = manifest?.modules;
    logger('manifest modules', apps);
  }

  // get module list
  if (manifest?.getAppList) {
    apps = await manifest?.getAppList();
    logger('getAppList modules', apps);
  }

  // get inject modules list
  if (window?.modern_manifest?.modules) {
    apps = window?.modern_manifest?.modules;
    logger('modern_manifest', apps);
  }

  return {
    apps,
    ...options,
  };
}

export default ((config: Config) => {
  setExternal();
  const { manifest = {}, LoadingComponent, ...options } = config;
  logger('createPlugin', { config });
  if (!manifest.LoadingComponent && LoadingComponent) {
    manifest.LoadingComponent = LoadingComponent;
  }

  const promise = initOptions(manifest, options);

  return createPlugin(() => ({
    hoc({ App }, next) {
      class GetMicroFrontendApp extends React.Component {
        state: {
          MApp: React.FC<MicroComponentProps>;
          apps: AppMap;
          appInfoList: ModulesInfo;
        } = {
          MApp: () => {
            logger('MApp init Component Render');
            return React.createElement('div');
          },
          apps: new Proxy(
            {},
            {
              get(_target, _p) {
                logger('apps init Component Render', _p);
                return () => React.createElement('div');
              },
            },
          ),
          appInfoList: [],
        };

        constructor(props: any) {
          super(props);
          const load = async () => {
            const GarfishConfig = await promise;
            logger('initOptions result', { manifest, GarfishConfig });
            const MApp = generateMApp(GarfishConfig, manifest);
            const { appInfoList, apps } = generateApps(GarfishConfig, manifest);
            logger('generateApps', { MApp, apps, appInfoList });
            this.setState({
              MApp,
              apps,
              appInfoList,
            });
          };
          load();
        }

        render() {
          logger('GarfishProvider state', this.state);
          return (
            <GarfishProvider value={this.state}>
              <App {...this.props} {...this.state} />
            </GarfishProvider>
          );
        }
      }

      return next({
        App: hoistNonReactStatics(GetMicroFrontendApp, App),
      });
    },
  }));
}) as any;
