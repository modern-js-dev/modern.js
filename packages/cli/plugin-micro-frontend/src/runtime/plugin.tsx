import { createPlugin } from '@modern-js/runtime-core';
import { useMemo } from 'react';
import hoistNonReactStatics from 'hoist-non-react-statics';
import { GarfishProvider } from './utils/Context';
import Apps from './utils/apps';
import setExternal from './utils/setExternal';
import { Config } from './typings';

export default ((config: Config) => {
  setExternal();
  const appsInstance = new Apps(config);

  return createPlugin(() => ({
    hoc({ App }, next) {
      appsInstance.run();

      const getMicroFrontendApp = (props: any) => {
        // eslint-disable-next-line react-hooks/rules-of-hooks
        const value = useMemo(
          () => ({
            MApp: appsInstance.getMApp(),
            apps: appsInstance.getApps(),
          }),
          [],
        );

        return (
          <GarfishProvider value={value}>
            <App {...props} />
          </GarfishProvider>
        );
      };
      return next({
        App: hoistNonReactStatics(getMicroFrontendApp, App),
      });
    },
  }));
}) as any;
