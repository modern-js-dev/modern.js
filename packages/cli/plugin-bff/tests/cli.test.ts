import path from 'path';
import { manager, AppContext } from '@modern-js/core';
import { modifyServerRoutes } from '@modern-js/plugin-analyze';
import Chain from 'webpack-chain';
import { CHAIN_ID } from '@modern-js/utils';
import plugin from '../src/cli';
import './helper';

describe('bff cli plugin', () => {
  it('schema', async () => {
    const main = manager.clone().usePlugin(plugin);
    const runner = await main.init();
    const result = await runner.validateSchema();

    expect(result).toMatchSnapshot();
  });

  it('config', async () => {
    const main = manager.clone().usePlugin(plugin);
    const runner = await main.init();
    const result = await runner.config();

    expect(result).toMatchSnapshot();
  });

  it('config webpack chain', async () => {
    const main = manager.clone().usePlugin(plugin);
    const runner = await main.init();
    const [{ tools }]: any = await runner.config();
    const chain = new Chain();
    AppContext.set({
      appDirectory: './fixtures/function',
      port: 3000,
    } as any);
    tools.webpackChain(chain, { CHAIN_ID });

    expect(chain.toConfig()).toMatchObject({
      module: {
        rules: [
          {
            oneOf: [
              {
                test: /.\/fixtures\/function\/api\/\.*(\.[tj]s)$/,
                use: [
                  {
                    loader: require
                      .resolve('../src/loader.ts')
                      .replace(/\\/g, '/'),
                    options: {
                      apiDir: path.resolve('./fixtures/function/api'),
                      fetcher: undefined,
                      port: 3000,
                      prefix: '/api',
                      target: undefined,
                    },
                  },
                ],
              },
            ],
          },
        ],
      },
      resolve: {
        alias: {
          '@api': path.resolve('./fixtures/function/api'),
        },
      },
    });
  });

  it('server routes', async () => {
    const main = manager.clone().usePlugin(plugin);
    main.registerHook({ modifyServerRoutes });
    const runner = await main.init();
    const result = await runner.modifyServerRoutes({ routes: [] });

    expect(result).toMatchSnapshot();
  });
});
