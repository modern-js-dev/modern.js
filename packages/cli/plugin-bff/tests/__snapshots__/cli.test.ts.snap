// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`bff cli plugin config 1`] = `
Array [
  Object {
    "source": Object {
      "moduleScopes": Array [
        "./api",
      ],
    },
    "tools": Object {
      "webpack": [Function],
    },
  },
]
`;

exports[`bff cli plugin config webpack chain 1`] = `
Object {
  "module": Object {
    "rules": Array [
      Object {
        "oneOf": Array [
          Object {
            "test": /\\.\\\\/fixtures\\\\/function\\\\/api\\\\/\\.\\*\\(\\.\\[tj\\]s\\)\\$/,
            "use": Array [
              Object {
                "loader": "/packages/cli/plugin-bff/src/loader.ts",
                "options": Object {
                  "apiDir": "/packages/cli/plugin-bff/fixtures/function/api",
                  "fetcher": undefined,
                  "port": 3000,
                  "prefix": "/api",
                  "target": undefined,
                },
              },
            ],
          },
        ],
      },
    ],
  },
  "resolve": Object {
    "alias": Object {
      "@api": "/packages/cli/plugin-bff/fixtures/function/api",
    },
  },
}
`;

exports[`bff cli plugin schema 1`] = `
Array [
  Array [
    Object {
      "schema": Object {
        "properties": Object {
          "fetcher": Object {
            "type": "string",
          },
          "prefix": Object {
            "type": "string",
          },
          "proxy": Object {
            "type": "object",
          },
          "requestCreater": Object {
            "type": "string",
          },
        },
        "type": "object",
      },
      "target": "bff",
    },
  ],
]
`;

exports[`bff cli plugin server routes 1`] = `
Object {
  "routes": Array [
    Object {
      "entryPath": "",
      "isApi": true,
      "isSPA": false,
      "isSSR": false,
      "urlPath": "/api",
    },
  ],
}
`;
